python/lsst/summit/utils/tmaUtils.py

```diff
@@ -0,0 +1,1588 @@
+        # preferable.
+        _initializeTma(tma)
+
+        tmaStates = {}
+        for rowNum, row in data.iterrows():
+            tma.apply(row)
+            tmaStates[rowNum] = tma.state
+
+        stateTuples = self._statesToEventTuples(tmaStates, dataIsForCurrentDay)
+        events = self._makeEventsFromStateTuples(stateTuples, dayObs, data)
+        self.addBlockDataToEvents(events)
+        return events
+
+    def _statesToEventTuples(self, states, dataIsForCurrentDay):
+        """Get the event-tuples from the dictionary of TMAStates.
+
+        Chunks the states into blocks of the same state, so that we can create
+        an event for each block in `_makeEventsFromStateTuples`. Off-type
+        states are skipped over, with each event starting when the telescope
+        next resumes motion or changes to a different type of motion state,
+        i.e. from non-tracking type movement (MOVE_POINT_TO_POINT, JOGGING,
+        TRACKING-but-not-in-position, i.e. slewing) to a tracking type
+        movement, or vice versa.
+
+        Parameters
+        ----------
+        states : `dict` of `int` : `lsst.summit.utils.tmaUtils.TMAState`
+            The states of the TMA, keyed by row number.
+        dataIsForCurrentDay : `bool`
+            Whether the data is for the current day. Determines whether to
+            allow and open last event or not.
+
+        Returns
+        -------
+        parsedStates : `list` of `tuple`
+            The parsed states, as a list of tuples of the form:
+                ``(eventStart, eventEnd, eventType, endReason)``
+        """
+        # Consider rewriting this with states as a list and using pop(0)?
+        skipStates = (TMAState.STOPPED, TMAState.OFF, TMAState.FAULT)
+
+        parsedStates = []
+        eventStart = None
+        rowNum = 0
+        nRows = len(states)
+        while rowNum < nRows:
+            previousState = None
+            state = states[rowNum]
+            # if we're not in an event, fast forward through off-like rows
+            # until a new event starts
+            if eventStart is None and state in skipStates:
+                rowNum += 1
+                continue
+
+            # we've started a new event, so walk through it and find the end
+            eventStart = rowNum
+            previousState = state
+            rowNum += 1  # move to the next row before starting the while loop
+            if rowNum == nRows:
+                # we've reached the end of the data, and we're still in an
+                # event, so don't return this presumably in-progress event
+                self.log.warning('Reached the end of the data while starting a new event')
+                break
+            state = states[rowNum]
+            while state == previousState:
+                rowNum += 1
+                if rowNum == nRows:
+                    break
+                state = states[rowNum]
```

Comment:
I see what you're getting at here, but I tried it, and I couldn't see how to do this in a way which didn't make the code look a lot more complex, rather than less. I'm very much up for tidying this part of the code up on a future ticket, because I really do agree with you and Russell both - this part is gross.